src/util/pcp.cpp

```diff
@@ -0,0 +1,520 @@
+        struct in_addr addr4;
+        std::memcpy(&addr4, wrapped_addr.data() + IPV4_IN_IPV6_PREFIX.size(), ADDR_IPV4_SIZE);
+        return CNetAddr(addr4);
+    } else {
+        struct in6_addr addr6;
+        std::memcpy(&addr6, wrapped_addr.data(), ADDR_IPV6_SIZE);
+        return CNetAddr(addr6);
+    }
+}
+
+//! PCP or NAT-PMP send-receive loop.
+static std::optional<std::vector<uint8_t>> PCPSendRecv(Sock &sock, const std::string &protocol, Span<const uint8_t> request, int num_tries,
+        std::chrono::milliseconds timeout_per_try,
+        std::function<bool(const Span<const uint8_t>)> check_packet)
+{
+    // UDP is a potentially lossy protocol, so we try to send again a few times.
+    uint8_t response[PCP_MAX_SIZE];
+    bool got_response = false;
+    int recvsz = 0;
+    for (int ntry = 0; !got_response && ntry < num_tries; ++ntry) {
+        if (ntry > 0) {
+            LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "%s: Retrying (%d)\n", protocol, ntry);
+        }
+        // Dispatch packet to gateway.
+        if (sock.Send(request.data(), request.size(), 0) != static_cast<ssize_t>(request.size())) {
+            LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "%s: Could not send request: %s\n", protocol, NetworkErrorString(WSAGetLastError()));
+            return std::nullopt; // Network-level error, probably no use retrying.
+        }
+
+        // Wait for response(s) until we get a valid response, a network error, or time out.
+        auto cur_time = std::chrono::time_point_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now());
+        auto deadline = cur_time + timeout_per_try;
+        while ((cur_time = std::chrono::time_point_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now())) < deadline) {
+            Sock::Event occurred = 0;
+            if (!sock.Wait(deadline - cur_time, Sock::RECV, &occurred)) {
+                LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "%s: Could not wait on socket: %s\n", protocol, NetworkErrorString(WSAGetLastError()));
+                return std::nullopt; // Network-level error, probably no use retrying.
+            }
+            if (!occurred) {
+                LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "%s: Timeout\n", protocol);
+                break; // Retry.
+            }
+
+            // Receive response.
+            recvsz = sock.Recv(response, sizeof(response), MSG_DONTWAIT);
+            if (recvsz < 0) {
+                LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "%s: Could not receive response: %s\n", protocol, NetworkErrorString(WSAGetLastError()));
+                return std::nullopt; // Network-level error, probably no use retrying.
+            }
+            LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "%s: Received response of %d bytes: %s\n", protocol, recvsz, HexStr(Span(response, recvsz)));
+
+            if (check_packet(Span<uint8_t>(response, recvsz))) {
```

Comment:
Unfortunately, that gives a new warning, because the `int` isn't the exact type:
```
../../src/util/pcp.cpp:258:41: warning: narrowing conversion of ‘recvsz’ from ‘int’ to ‘std::size_t’ {aka ‘long unsigned int’} [-Wnarrowing]
  258 |             if (check_packet({response, recvsz})) {
```
Could add a `static_cast` but i don't think it's much of an improvement in that case?